@ktf @jgrosseo ok, with https://github.com/AliceO2Group/AliceO2/pull/9715 we can have this:
```
root [0] auto& cm = o2::ccdb::BasicCCDBManager::instance();
[INFO] Init CcdApi with UserAgentID: alicrs02-1661771661-pnSfKI, Host: http://alice-ccdb.cern.ch/
root [1] cm.setFatalWhenNull(false);
root [2] cm.get<o2::parameters::GRPECSObject>("GLO/Config/GRPECS");
root [3] cm.get<std::vector<o2::detectors::AlignParam>>("ITS/Calib/Align");
root [4] cm.get<std::vector<o2::detectors::AlignParam>>("ITS/Calib/Align");
root [5] cm.get<std::vector<o2::detectors::AlignParam>>("ITS/Calib/Bla");
[ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch//ITS/Calib/Bla/1661771661006/
root [6] cm.get<std::vector<o2::detectors::AlignParam>>("ITS/Calib/Bla");
[ERROR] Requested resource does not exist: http://alice-ccdb.cern.ch//ITS/Calib/Bla/1661771661006/
root [7] cm.endOfStream();
[INFO] 5 queries for 3 objects, 2 good fetches (and 2 failed ones for 1 objects), instance: alicrs02-1661771661-pnSfKI
```
Is this enough or do you want more detailed output?

--- a/Framework/Core/include/Framework/AnalysisManagers.h
+++ b/Framework/Core/include/Framework/AnalysisManagers.h
@@ -413,6 +413,26 @@ class has_instance
 };
 
 template <typename T>
+class has_end_of_stream
+{
+  typedef char one;
+  struct two {
+    char x[2];
+  };
+
+  template <typename C>
+  static one test(decltype(&C::endOfStream));
+  template <typename C>
+  static two test(...);
+
+ public:
+  enum { value = sizeof(test<T>(nullptr)) == sizeof(char) };
+};
+
+template <typename T>
+inline constexpr bool has_end_of_stream_v = has_end_of_stream<T>::value;
+
+template <typename T>
 struct ServiceManager {
   template <typename ANY>
   static bool add(std::vector<ServiceSpec>&, ANY&)
@@ -422,6 +442,12 @@ struct ServiceManager {
 
   template <typename ANY>
   static bool prepare(InitContext&, ANY&)
+  {
+    return false;
+  }
+
+  template <typename ANY>
+  static bool postRun(EndOfStreamContext&, ANY&)
   {
     return false;
   }
@@ -442,6 +468,18 @@ struct ServiceManager<Service<T>> {
       return true;
     } else {
       service.service = &(context.services().get<T>());
+      return true;
+    }
+    return false;
+  }
+
+  /// If a service has a method endOfStream, it is called at the end of the stream.
+  static bool postRun(EndOfStreamContext& context, Service<T>& service)
+  {
+    // FIXME: for the moment we only need endOfStream to be
+    // stateless. In the future we might want to pass it EndOfStreamContext
+    if constexpr (has_end_of_stream_v<T>) {
+      service.service->endOfStream();
       return true;
     }
     return false;

--- a/Framework/Core/include/Framework/AnalysisTask.h
+++ b/Framework/Core/include/Framework/AnalysisTask.h
@@ -605,7 +605,11 @@ DataProcessorSpec adaptAnalysisTask(ConfigContext const& ctx, Args&&... args)
 
     auto& callbacks = ic.services().get<CallbackService>();
     auto endofdatacb = [task](EndOfStreamContext& eosContext) {
-      homogeneous_apply_refs([&eosContext](auto&& x) { return OutputManager<std::decay_t<decltype(x)>>::postRun(eosContext, x); }, *task.get());
+      homogeneous_apply_refs([&eosContext](auto&& x) {
+          using X = std::decay_t<decltype(x)>;
+          ServiceManager<X>::postRun(eosContext, x);
+          return OutputManager<X>::postRun(eosContext, x); },
+                             *task.get());
       eosContext.services().get<ControlService>().readyToQuit(QuitRequest::Me);
     };
     callbacks.set(CallbackService::Id::EndOfStream, endofdatacb);

